Branch-SDK/Branch-SDK/BNCServerInterface.m

```diff
@@ -14,14 +14,35 @@
 @implementation BNCServerInterface
 
 // make a generalized get request
-- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag {
-    [self getRequestAsync:params url:url andTag:requestTag log:YES];
+- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag retryCount:(NSInteger)retryCount {
+    [self getRequestAsync:params url:url andTag:requestTag retryNumber:0 retryCount:retryCount log:YES];
 }
 
 // this is actually a synchronous call; it should NOT be called from the main queue
-- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag log:(BOOL)log {
-    BNCServerResponse *serverResponse = [self genericSyncHTTPRequest:[self prepareGetRequest:params url:url log:log] withTag:requestTag andLinkData:nil];
-    if (self.delegate) [self.delegate serverCallback:serverResponse];
+- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag retryCount:(NSInteger)retryCount log:(BOOL)log {
+    [self getRequestAsync:params url:url andTag:requestTag retryNumber:0 retryCount:retryCount log:log];
+}
+
+- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag retryNumber:(NSInteger)retryNumber retryCount:(NSInteger)retryCount log:(BOOL)log {
+    BNCServerResponse *serverResponse = [self genericSyncHTTPRequest:[self prepareGetRequest:params url:url retryNumber:retryNumber log:log] withTag:requestTag andLinkData:nil];
+
+    NSInteger status = [serverResponse.statusCode integerValue];
+    BOOL isRetryableStatusCode = status >= 500;
+    
+    // Retry the request if appropriate
+    if (retryCount > 0 && isRetryableStatusCode) {
+        [NSThread sleepForTimeInterval:[BNCPreferenceHelper getRetryInterval]];
+
+        if (log) {
+            [BNCPreferenceHelper log:FILE_NAME line:LINE_NUM message:@"Replaying request with tag %@", requestTag];
+        }
+
+        [self getRequestAsync:params url:url andTag:requestTag retryNumber:++retryNumber retryCount:--retryCount log:log];
```

Comment:
Actually, yeah. I was thinking about how to do that, and just now realized a better solution. Will refactor with that shortly.